src/modules/0.0.18/aws_cloudfront/index.ts

```diff
@@ -0,0 +1,288 @@
+      create: async (es: Distribution[], ctx: Context) => {
+        const client = await ctx.getAwsClient() as AWS;
+        const out = [];
+        for (const e of es) {
+          const config:DistributionConfig = {
+            CallerReference: e.callerReference,
+            Comment: e.comment,
+            Enabled: e.enabled,
+            IsIPV6Enabled: e.isIPV6Enabled,
+            WebACLId: e.webACLId,
+            Origins: { Quantity: e.origins.length, Items: e.origins},
+            DefaultCacheBehavior: e.defaultCacheBehavior,
+          };
+          const res = await this.createDistribution(
+            client.cloudfrontClient, {
+              DistributionConfig: config
+            }
+          );
+
+          if (res) {
+            await waitUntilDistributionDeployed({
+              client: client.cloudfrontClient,
+              // all in seconds
+              maxWaitTime: 900,
+              minDelay: 1,
+              maxDelay: 4,
+            } as WaiterOptions<CloudFront>, { Id: res.Distribution?.Id, });
+
+            if (res && res.Distribution) {
+              const newDistribution = this.distributionMapper(res);
+              newDistribution.id = e.id;
+              newDistribution.location = res.Location;
+              newDistribution.status = "Deployed";
+              await this.module.distribution.db.update(newDistribution, ctx);
+              out.push(newDistribution);
+            }
+          }
+        }
+        return out;
+      },
+      read: async (ctx: Context, id?: string) => {
+        const client = await ctx.getAwsClient() as AWS;
+        if (id) {
+          const rawDistribution = await this.getDistribution(client.cloudfrontClient, id);
+          if (rawDistribution) {
+            const result = this.distributionMapper(rawDistribution);
+            return result;
+          }
+        } else {
+          const distributions = await this.getDistributions(client.cloudfrontClient);
+          const out = [];
+          for (const distribution of distributions) {
+            out.push(this.distributionMapper(distribution));
+          }
+          return out;
+        }
+      },
+      updateOrReplace: () => 'update',
+      update: async (es: Distribution[], ctx: Context) => {
+        const client = (await ctx.getAwsClient()) as AWS;
+        const out = [];
+        for (const e of es) {
+          if (e.status!=="Deployed") continue; // do not modify until it is deployed
+          const cloudRecord = ctx?.memo?.cloud?.Distribution?.[e.distributionId ?? ""];
+          const isUpdate = Object.is(
+            this.module.distribution.cloud.updateOrReplace(cloudRecord, e),
+            'update'
+          );
+          if (isUpdate) {
+            if (!e.eTag) throw new Error("Cannot update a distribution without an etag");  // cannot update without etag
+            const req = await this.getDistributionConfigForUpdate(client.cloudfrontClient, e);
+
+            if (e.eTag) {
```

Comment:
nit: I think this is not necessary since we are already `throw`ing if `!e.eTag` in line 225